File: src/main/java/engineer/carrot/warren/thump/command/minecraft/CommandThump.java
package engineer.carrot.warren.thump.command.minecraft;

import com.google.common.collect.Iterables;
import com.google.common.collect.Lists;
import com.google.common.collect.Maps;
import engineer.carrot.warren.thump.command.minecraft.handler.*;
import engineer.carrot.warren.thump.connection.ConnectionManager;
import engineer.carrot.warren.thump.util.helper.PredicateHelper;
import joptsimple.internal.Strings;
import net.minecraft.command.CommandBase;
import net.minecraft.command.ICommandSender;
import net.minecraft.util.ChatComponentText;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

public class CommandThump extends CommandBase {
    private ConnectionManager manager;
    private Map<String, ICommandHandler> handlers;

    private static final String COMMAND_NAME = "thump";
    private static final String COMMAND_USAGE = "";

    public CommandThump(ConnectionManager manager) {
        this.manager = manager;

        this.handlers = Maps.newHashMap();
        this.handlers.put("status", new StatusCommandHandler(manager));
        this.handlers.put("connect", new ConnectCommandHandler(manager));
        this.handlers.put("disconnect", new DisconnectCommandHandler(manager));
        this.handlers.put("reload", new ReloadCommandHandler(manager));
    }

    // CommandBase

    @Override
    public String getCommandName() {
        return COMMAND_NAME;
    }

    @Override
    public String getCommandUsage(ICommandSender sender) {
        return "/" + this.getCommandName() + " " + Strings.join(Lists.newArrayList(this.handlers.keySet()), ", ");
    }

    @Override
    public void processCommand(ICommandSender sender, String[] parameters) {
        if (parameters.length < 1 || !this.handlers.containsKey(parameters[0])) {
            sender.addChatMessage(new ChatComponentText("Invalid usage."));
            sender.addChatMessage(new ChatComponentText(" Usage: " + this.getCommandUsage(sender)));

            return;
        }

        this.handlers.get(parameters[0]).processParameters(sender, Arrays.copyOfRange(parameters, 1, parameters.length));
    }

    @Override
    public List<String> addTabCompletionOptions(ICommandSender sender, String[] parameters) {
        if (parameters.length <= 1) {
            String handlerId = (parameters[0] == null) ? "" : parameters[0];
            return Lists.newArrayList(Iterables.filter(
                    this.handlers.keySet(),
                    new PredicateHelper.StartsWithPredicate(handlerId)
            ));
        }

        String handlerId = parameters[0];
        if (!this.handlers.containsKey(handlerId)) {
            return null;
        }

        return this.handlers.get(handlerId).addTabCompletionOptions(sender, Arrays.copyOfRange(parameters, 1, parameters.length));
    }
}

File: src/main/kotlin/engineer/carrot/warren/thump/command/minecraft/CommandThump.kt
package engineer.carrot.warren.thump.command.minecraft

import com.google.common.base.Joiner
import engineer.carrot.warren.thump.api.ICommandHandler
import engineer.carrot.warren.thump.command.minecraft.handler.ReloadCommandHandler
import engineer.carrot.warren.thump.command.minecraft.handler.StatusCommandHandler
import engineer.carrot.warren.thump.plugin.IThumpServicePlugins
import net.minecraft.command.CommandBase
import net.minecraft.command.ICommandSender
import net.minecraft.server.MinecraftServer
import net.minecraft.util.math.BlockPos
import net.minecraft.util.text.TextComponentString

class CommandThump(private val servicePlugins: IThumpServicePlugins) : CommandBase() {

    private val handlers = mutableMapOf<String, ICommandHandler>()
    private val serviceHandlers = mutableMapOf<String, ICommandHandler>()

    fun reconfigureHandlers(pluginCommandHandlers: Map<String, ICommandHandler>) {
        handlers.clear()
        serviceHandlers.clear()

        handlers.put("reload", ReloadCommandHandler(servicePlugins))
        handlers.put("status", StatusCommandHandler(servicePlugins))

        serviceHandlers.putAll(pluginCommandHandlers.filterNot { handlers.containsKey(it.key) })

        handlers.put("service", object : ICommandHandler {
            override val command = "service"
            override val usage = "$command ${serviceHandlers.keys.joinToString(separator = ", ")}"

            override fun processParameters(sender: ICommandSender, parameters: Array<String>) {
                if (parameters.isEmpty() || !serviceHandlers.containsKey(parameters[0])) {
                    sender.sendMessage(TextComponentString("Invalid usage."))
                    sender.sendMessage(TextComponentString(" Usage: $usage"))
                } else {
                    serviceHandlers[parameters[0]]?.processParameters(sender, parameters.drop(1).toTypedArray())
                }
            }

            override fun addTabCompletionOptions(sender: ICommandSender, parameters: Array<String>): List<String> {
                return when (parameters.size) {
                    0 -> serviceHandlers.keys.toList()
                    1 -> serviceHandlers.keys.filter { it.startsWith(parameters[0]) }
                    else -> serviceHandlers[parameters[0]]?.addTabCompletionOptions(sender, parameters.drop(1).toTypedArray()) ?: listOf()
                }
            }
        })
    }

    // CommandBase

    override fun getName(): String {
        return COMMAND_NAME
    }

    override fun getUsage(sender: ICommandSender): String {
        return "/" + this.name + " " + Joiner.on(", ").join(this.handlers.keys)
    }

    override fun execute(server: MinecraftServer?, sender: ICommandSender?, parameters: Array<out String>?) {
        if (sender == null || parameters == null) {
            return
        }

        if (parameters.isEmpty() || !this.handlers.containsKey(parameters[0])) {
            sender.sendMessage(TextComponentString("Invalid usage."))
            sender.sendMessage(TextComponentString(" Usage: " + this.getUsage(sender)))

            return
        }

        this.handlers[parameters[0]]?.processParameters(sender, parameters.drop(1).toTypedArray())
    }

    override fun getTabCompletions(server: MinecraftServer?, sender: ICommandSender?, args: Array<out String>?, targetPos: BlockPos?): MutableList<String> {
        if (args == null) {
            return mutableListOf()
        }

        if (sender == null) {
            return mutableListOf()
        }

        return when (args.size) {
            0 -> handlers.keys.toMutableList()
            1 -> handlers.keys.filter { it.startsWith(args[0]) }.toMutableList()
            else -> handlers[args[0]]?.addTabCompletionOptions(sender, args.drop(1).toTypedArray())?.toMutableList() ?: mutableListOf()
        }
    }

    override fun getRequiredPermissionLevel(): Int {
        return COMMAND_DEFAULT_PERMISSION_LEVEL
    }

    companion object {
        private val COMMAND_NAME = "thump"
        private val COMMAND_DEFAULT_PERMISSION_LEVEL = 2
    }
}
